packages/shorebird_cli/lib/src/config/shorebird_yaml.dart

```diff
@@ -13,6 +13,19 @@ class ShorebirdYaml {
   factory ShorebirdYaml.fromJson(Map<dynamic, dynamic> json) =>
       _$ShorebirdYamlFromJson(json);
 
-  final String appId;
+  @JsonKey(fromJson: AppId.fromJson)
+  final AppId appId;
   final String? baseUrl;
 }
+
+class AppId {
+  const AppId({this.value, this.values});
+
+  factory AppId.fromJson(dynamic json) {
+    if (json is String) return AppId(value: json);
+    return AppId(values: (json as Map).cast<String, String>());
+  }
+
+  final String? value;
+  final Map<String, String>? values;
```

Comment:
Given that only one of these can be valid, I worry that the code to handle this could get kind of boilerplate-y. I don't have any awesome suggestions on how to fix (I want to reach for Swift's [enums with associated types](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/enumerations/#Associated-Values), but I don't think an analog to that exists in Dart).

Reply:
Yeah I have some ideas in mind but wanted to hold off introducing more code until it's needed.